### ISSUE TITLE
[FrameworkBundle] [Cache] cache:clear warms up cache incorrectly

### ISSUE THREAD

**Issue by benbor** [symfony/symfony#26219]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | yes
| RFC?             | no
| Symfony version  | >4.0.3

After update Symfony to version 4.0.3 (4.0.4 -the same)  cache:clear command (with implicit cache warming) warms up cache incorrectly. As i can compare with 4.0.2 cache folder some dependencies is not warmed. 

**4.0.3**
```bash
root@bacac0721dd9:/var/www/symfony# bin/console ca:cl -e=prod
root@bacac0721dd9:/var/www/symfony# ls -1 var/cache/prod/
ContainerM4N54hq
annotations.map
srcProdProjectContainer.php
srcProdProjectContainer.php.meta
srcProdProjectContainerUrlGenerator.php
srcProdProjectContainerUrlGenerator.php.meta
srcProdProjectContainerUrlMatcher.php
srcProdProjectContainerUrlMatcher.php.meta
templates.php
```

**4.0.2**
```bash
root@bacac0721dd9:/var/www/symfony# bin/console ca:cl -e=prod
root@bacac0721dd9:/var/www/symfony# ls -1 var/cache/prod/
ContainerW99Fklq
annotations.map
annotations.php
doctrine
easy_admin
pools
serialization.php
srcProdProjectContainer.php
srcProdProjectContainer.php.meta
srcProdProjectContainerUrlGenerator.php
srcProdProjectContainerUrlGenerator.php.meta
srcProdProjectContainerUrlMatcher.php
srcProdProjectContainerUrlMatcher.php.meta
templates.php
translations
twig
validation.php
```

For develop env it isn't critical, but for production it is.
ps As workaround i used `bin/console cache:warmup -e=prod` after

**Comment by xabbuh** [comment]
Can you create a small example app that allows to reproduce your issue?

**Comment by benbor** [comment]
Hi @xabbuh. 

I've tried to reproduce issue on small example (framework + templating + doctrine). The bug isn't reproduced. 
Looks like some another third party bundle  in my project failed while warmed and console command just silently stopped. I'm think this behavior was introduced with #25117. 
While i was trying play with composer dependencies broken lib was updated and fixed, and bug no more reproduced.

**Comment by stof** [comment]
it would still be great to reproduce this to at least make the failure not silent.

**Comment by lushc** [comment]
Hmm, looking into this some more, I think I've got an idea of what's going on. The crux of the issue is that the cache warmer is being passed the cache dir (`prod`) instead of the warmup dir (`pro_`), so warmup cache is generated in the wrong directory. However, this only happens when the warmup container dumps with the same content hash as the previously cached container in `prod`. Because the `appProdProjectContainer` classes are generated with the same namespace, e.g. `ContainerBnhddds`, but rely on their current directory to generate the right target dirs, an instance of the container in the `prod` folder is returned which gives a `kernel.cache_dir` value mis-match.

1. Kernel boots for the first time when running the `cache:clear` command.
2. `var/cache/prod/appProdProjectContainer.php` is included in `Kernel::initializeContainer`, which returns a new instance of `\ContainerBnhddds\appProdProjectContainer`.
3. Later on, the kernel is rebooted with `CacheClearCommand::warmup`, which compiles the container as the warmup cache isn't fresh (`var/cache/pro_/appProdProjectContainer.php` doesn't yet exist).
4. The new container has the same build ID, so while `var/cache/pro_/appProdProjectContainer.php` is required, an instance of the `\ContainerBnhddds\appProdProjectContainer` class in `prod` is returned.
5. The cache warmer now generates cache in `prod`, but since `pro_` is later renamed to `prod` in `CacheClearCommand::execute`, any generated files are lost.

I hope this makes sense, I've been debugging for a good few hours since I noticed that one 3.4 project exhibited this behaviour and another didn't, which lead me to #761 and this issue. I'm not sure why only one project always generates a different build ID each time, or if it's even meant to, but this clash is certainly causing warmup cache to be put in the wrong directory.

**Comment by benbor** [comment]
@stof I debugged it for 4.0.5
I would like to authoritatively say that @lushc is totally right. 
Unfortunately #25958 doesn't fix it.

**Comment by benbor** [comment]
I need to summon author of  "reproducible builds" to clarify this issue.
@lstrojny please could you help us.

**Comment by lstrojny** [comment]
@benbor trying to understand how @lushc explanation could be triggered by the changes implemented for #25958 but I am lost. 
#26176 and #26128 are the two container related changes. I don’t see how these changes could introduce the behavior illustrated here.

**Comment by benbor** [comment]
@lstrojny I'm sorry.
Exactly  #25958, #26176 and #26128 changes doesn't trigger this issue.
But this issue happens, because of "reproducible builds". As long as container doesn't changed new generated class has the same name as old container class, so PHP can't load it. 
@lstrojny i would ask you to help us resolve this problem, because you have experience with reproducible build.
I have one idea. 
I can add $time to container name. For example https://github.com/symfony/dependency-injection/blob/master/Dumper/PhpDumper.php#L231-L245
will looks like:
```php
            $code[$options['class'].'.php'] = <<<EOF
<?php
{$namespaceLine}
// This file has been auto-generated by the Symfony Dependency Injection Component for internal use.
if (\\class_exists(\\Container{$hash}{$time}\\{$options['class']}::class, false)) {
    // no-op
} elseif (!include __DIR__.'/Container{$hash}{$time}/{$options['class']}.php') {
    touch(__DIR__.'/Container{$hash}{$time}.legacy');
    return;
}
if (!\\class_exists({$options['class']}::class, false)) {
    \\class_alias(\\Container{$hash}{$time}\\{$options['class']}::class, {$options['class']}::class, false);
}
return new \\Container{$hash}{$time}\\{$options['class']}(array(
    'container.build_hash' => '$hash',
    'container.build_id' => '$id',
    'container.build_time' => $time,
));
EOF;
```
I could create PR, if you agree with this solution.
Feel free to propose better solution. Thanks

**Comment by lstrojny** [comment]
@benbor adding the time would actually reverse the effort with reproducible builds, that’s not a good idea.
Isn’t the problem in the cache warmer because the wrong path is passed?

**Comment by benbor** [comment]
@lstrojny excuse me, but after #26128 the time is optionally deterministic. 
For guys who doesn't care about reproducible builds (time will difference each run) solution will works fine. `bin/console ca:cl -e=prod` will works correct.
Guys who care about reproducible builds, have to  fix the build time in parameters, so they will have the same container name (unfortunately, `bin/console ca:cl -e=prod` will works incorrect, the same as at now :( )

> Isn’t the problem in the cache warmer because the wrong path is passed?

roughly speaking  - yes. 
I  try explain a bit. When `bin/console ca:cl -e=prod` started, before command executing  Symfony create and dump old kernel in `/var/cache/prod/srcProdProjectContainer.php`, and then load  class via `new ContainerA9d1Z7x\srcProdProjectContainer` into PHP memory.
After that CacheClearCommand   tried to load `/var/cache/pro_/srcProdProjectContainer.php` witch create new instance of `/var/cache/pro_/ContainerA9d1Z7x/srcProdProjectContainer.php` , but  PHP already has \ContainerA9d1Z7x\srcProdProjectContainer (from `/var/cache/prod` path) and return it.

So, i think, as PHP can not unload classes from memory, we cant resolve this  issue without changing class name.

**Comment by lstrojny** [comment]
@benbor I get it now, thanks for your patience. Since we can fixate the time for optional builds it should indeed work, as you proposed, to include the time in the class name and be done with it.

**Comment by VincentFoulon80** [comment]
I am actually encountering the same issue when attempting to migrate my project from Symfony 3.3 to 3.4 (tried 3.4.0 and 3.4.6).
I have a simple action that clears the cache:
```php
public function cacheClearAction()
    {
        $fs = new Filesystem();
        $fs->remove($this->getParameter('kernel.cache_dir'));
        return new JsonResponse(true);
    }
```
The cache directory is cleared, but Symfony gives me this error instead of returning the response:

> Compile Error: ContainerNeqkqxx\appDevDebugProjectContainer::load(): Failed opening required '/{mypath}/var/cache/dev/ContainerNeqkqxx/getSecurity_Access_AuthenticatedVoterService.php'
>in appDevDebugProjectContainer.php (line 923)

I don't know if this is related to the current issue, but when reading the messages it just made me think of it.

**Comment by fancyweb** [comment]
@VincentFoulon80 I think your issue is more related to https://github.com/symfony/symfony/issues/25654

**Comment by VincentFoulon80** [comment]
@fancyweb You are right, I didn't see that issue when searching for my problem. My bad. Thanks by the way.

**Comment by Tobion** [comment]
This looks like a duplicate of #26506. Please check if #26662 fixes the issue.

**Comment by blackandred** [comment]
I have also the issue.

```
[2018-03-28 11:51:18] php.WARNING: Warning: require(/var/www/var/cache/prod/ContainerAGKUHwl/getPasswordUserChangedNotificationService.php): failed to open stream: No such file or directory {"exception":"[object] (ErrorException(code: 0): Warning: require(/var/www/var/cache/prod/ContainerAGKUHwl/getPasswordUserChangedNotificationService.php): failed to open stream: No such file or directory at /var/www/var/cache/prod/ContainerAGKUHwl/srcProdProjectContainer.php:307)"} []
[2018-03-28 11:51:18] php.CRITICAL: Fatal Compile Error: require(): Failed opening required '/var/www/var/cache/prod/ContainerAGKUHwl/getPasswordUserChangedNotificationService.php' (include_path='/var/www/src/application/site/includes:/var/www/vendor/pear/console_getopt:/var/www/vendor/pear/db:/var/www/vendor/pear/file:/var/www/vendor/pear/pear-core-minimal/src:/var/www/vendor/pear/pear_exception:.:/usr/local/lib/php:/var/www/src/application/site') {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalErrorException(code: 0): Compile Error: require(): Failed opening required '/var/www/var/cache/prod/ContainerAGKUHwl/getPasswordUserChangedNotificationService.php' (include_path='/var/www/src/application/site/includes:/var/www/vendor/pear/console_getopt:/var/www/vendor/pear/db:/var/www/vendor/pear/file:/var/www/vendor/pear/pear-core-minimal/src:/var/www/vendor/pear/pear_exception:.:/usr/local/lib/php:/var/www/src/application/site') at /var/www/var/cache/prod/ContainerAGKUHwl/srcProdProjectContainer.php:307)"} []
```

The `./bin/console cache:clear --env=prod` and `./bin/console cache:warmup --env=prod` ends with success, but the container does not have a proper service warmed up.

Container definition:
```
    AppBundle\Components\ResettingPassword\Event\Subscriber\PasswordUserChangedNotification:
        class: AppBundle\Components\ResettingPassword\Event\Subscriber\PasswordUserChangedNotification
        arguments:
            $templating: '@AppBundle\Service\Templating'
            $mailer: '@AppBundle\Components\Mailing\Mailer'
            $translator: '@AppBundle\Components\Language\TranslationProvider'
            $appDomainUrl: '%env(APP_DOMAIN_URL)%'
```

This is tested on a test server, when I switch `APP_ENV` to `dev` then it works, so it must be about the static cache warmup.

This issue makes imposible to release the software.

**Comment by blackandred** [comment]
It seems that it may be related to opcache - I restarted the supervisor in the docker container and it works.
The issue may be that the `srcProdProjectContainer.php` is cached by the opcache.

Possibly in case of using the opcache setting a proper value for the `opcache.blacklist_filename` may solve the issue.

**Comment by benbor** [comment]
@blackandred look at my previously comment. 
In `dev` environment your application work, because in `dev` symfony can warm up services on the fly.
For test please try following command: `bin/console cache:clear --env=dev && ls -l var/cache/dev`
It show you all files in cache. After that open your application (or just do `bin/console cache:warmup --env=dev`) and compare cache file list  with previous one.

**Comment by nicolas-grekas** [comment]
As spotted by @Tobion:

> This looks like a duplicate of #26506. Please check if #26662 fixes the issue.

could anyone having the issue confirm?
